File: persistence/src/main/java/de/cyface/persistence/model/Measurement.java
package de.cyface.persistence.model;

import static de.cyface.persistence.FileUtils.generateMeasurementFilePath;
import static de.cyface.persistence.FileUtils.generateMeasurementFolderPath;

import java.io.File;
import java.io.IOException;

import android.content.Context;
import android.util.Log;
import androidx.annotation.NonNull;
import de.cyface.persistence.Constants;
import de.cyface.persistence.Persistence;
import de.cyface.persistence.serialization.MetaFile;
import de.cyface.utils.Validate;

/**
 * An object of this class represents a single measurement captured by the {@code DataCapturingService}. This usually
 * happens between complementary calls to
 * {@code DataCapturingService#startAsync(DataCapturingListener, Vehicle, StartUpFinishedHandler)} and
 * {@code DataCapturingService#stopAsync(ShutDownFinishedHandler)}.
 *
 * @author Klemens Muthmann
 * @author Armin Schnabel
 * @version 2.2.0
 * @since 1.0.0
 */
public final class Measurement {
    /**
     * The system wide unique identifier of this measurement.
     */
    private long id;
    /**
     * The {@link MeasurementStatus} of the measurement tells weather the measurement is still capturing data, finished
     * or similar.
     */
    private MeasurementStatus status;
    /**
     * The {@link Context} required to access the underlying persistence layer.
     */
    private Context context;
    /**
     * The file to persist meta information of a measurement.
     */
    private MetaFile metaFile;

    /**
     * Creates a new {@link Measurement}. Use {@link Persistence#newMeasurement(Vehicle)} instead to create a new
     * measurement it in the persistence layer, too.
     *
     * @param context The {@link Context} required to access the underlying persistence layer.
     * @param id The system wide unique identifier of this measurement.
     * @param status The {@link MeasurementStatus} of the measurement tells weather the measurement is still capturing
     *            data, finished or similar.
     */
    // TODO: we should check somewhere (when creating/deleting a m?) that there is only one folder per measurement id)
    public Measurement(@NonNull final Context context, final long id, @NonNull final MeasurementStatus status) {
        this.id = id;
        this.status = status;
        this.context = context;
    }

    /**
     * Returns a link to the folder containing the measurement data.
     *
     * @return The {@link File} link to the measurement's folder.
     */
    public File getMeasurementFolder() {
        return generateMeasurementFolderPath(context, status, id);
    }

    @Override
    public String toString() {
        return "Measurement{" + "id=" + id + ", status=" + status + ", context=" + context + ", metaFile=" + metaFile
                + '}';
    }

    // TODO how to handle when the MetaFile was not yet loaded? Maybe make sure that it's aways loaded?
    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        Measurement that = (Measurement)o;
        return id == that.id && status == that.status && context.equals(that.context)
                && ((metaFile == null && that.metaFile == null)
                        || (metaFile != null && that.metaFile != null && metaFile.equals(that.metaFile)));
    }

    @Override
    public int hashCode() {
        return (int)(id ^ (id >>> 32));
    }

    public long getIdentifier() {
        return id;
    }

    public MeasurementStatus getStatus() {
        return status;
    }

    /**
     *
     * Close the specified {@link Measurement}.
     * (!) Attention: This only moves the measurement to the finished folder. The
     * <code>DataCapturingBackgroundService</code>
     * must have stopped normally in advance in order for the point counts to be written into the {@link MetaFile}.
     * Else, the file is seen as corrupted.
     *
     * Supported MeasurementStatus life-cycle:
     * - {@link Persistence#newMeasurement(Vehicle)} -> {@link MeasurementStatus#OPEN}
     *
     * - {@link MeasurementStatus#OPEN} -> {@link MeasurementStatus#PAUSED}
     * - {@link MeasurementStatus#OPEN} -> {@link MeasurementStatus#FINISHED}
     *
     * - {@link MeasurementStatus#PAUSED} -> {@link MeasurementStatus#OPEN}
     * - {@link MeasurementStatus#PAUSED} -> {@link MeasurementStatus#FINISHED}
     *
     * - {@link MeasurementStatus#FINISHED} -> {@link MeasurementStatus#SYNCED}
     */
    public void setStatus(@NonNull final MeasurementStatus newStatus) {
        if (status == MeasurementStatus.SYNCED || status == MeasurementStatus.CORRUPTED) {
            throw new IllegalStateException("Unsupported MeasurementStatus life-cycle");
        }

        final boolean supportedOpenMeasurementLifeCycleFlow = status == MeasurementStatus.OPEN
                && (newStatus == MeasurementStatus.PAUSED || newStatus == MeasurementStatus.FINISHED
                        || newStatus == MeasurementStatus.CORRUPTED);
        final boolean supportedPausedMeasurementLifeCycleFlow = status == MeasurementStatus.PAUSED
                && (newStatus == MeasurementStatus.OPEN || newStatus == MeasurementStatus.FINISHED);
        final boolean supportedFinishedMeasurementLifeCycleFlow = status == MeasurementStatus.FINISHED
                && (newStatus == MeasurementStatus.SYNCED);

        if (supportedOpenMeasurementLifeCycleFlow || supportedPausedMeasurementLifeCycleFlow
                || supportedFinishedMeasurementLifeCycleFlow) {
            move(newStatus);
            status = newStatus;
            metaFile = MetaFile.loadFile(this); // Because the measurement was moved to another location
            return;
        }

        throw new IllegalStateException("Illegal MeasurementStatus life-cycle");

    }

    /**
     * Moves a {@link Measurement} from one {@link MeasurementStatus} folder to another status folder.
     *
     * @param newStatus the target status of the measurement to move.
     * @throws IllegalStateException when the current measurement folder does not exist or when the measurement could
     *             not be moved.
     */
    private void move(MeasurementStatus newStatus) {
        synchronized (this) {
            final File measurementFolder = getMeasurementFolder();
            if (!measurementFolder.exists()) {
                throw new IllegalStateException("Failed to access non existent measurement: " + measurementFolder);
            }

            final File targetMeasurementFolder = generateMeasurementFolderPath(context, newStatus, id);
            if (!measurementFolder.renameTo(targetMeasurementFolder)) {
                throw new IllegalStateException("Failed to move measurement: " + measurementFolder.getParent() + " to "
                        + targetMeasurementFolder.getPath());
            }
        }
    }

    /**
     * Creates the folder for this measurement. Its {@link MeasurementStatus} defined where the folder is created.
     */
    public void createDirectory() {
        final File measurementDir = getMeasurementFolder();
        if (measurementDir.exists()) {
            throw new IllegalStateException("Cannot create measurement folder as it already exists.");
        }
        if (!measurementDir.mkdirs()) {
            throw new IllegalStateException("Failed to create measurement folder: " + measurementDir.getAbsolutePath());
        }
    }

    /**
     * Creates a file for Cyface binary data for this measurement.
     *
     * @param fileName the name of the file
     * @param fileExtension the extension of the file
     * @return A {@link File} link to the created file.
     * @throws IllegalStateException when the measurement folder does not exist.
     */
    public File createFile(@NonNull final String fileName, @NonNull final String fileExtension) {
        final File file = generateMeasurementFilePath(this, fileName, fileExtension);
        Validate.isTrue(!file.exists(), "Failed to createFile as it already exists: " + file.getPath());
        try {
            if (!file.createNewFile()) {
                throw new IOException("Failed to createFile: " + file.getPath());
            }
            Validate.isTrue(file.exists());
            Log.d(Constants.TAG, "CreateFile successful: " + file.getAbsolutePath());
        } catch (final IOException e) {
            throw new IllegalStateException("Failed createFile: " + file.getPath());
        }
        return file;
    }

    public MetaFile getMetaFile() {
        return metaFile;
    }

    /**
     * Loads the {@link MetaFile} from the persistence layer and re-links it to this measurement internally.
     */
    public void loadMetaFile() {
        Validate.isTrue(metaFile == null, "Wrong method use: MetaFile already linked.");
        metaFile = MetaFile.loadFile(this);
        Validate.isTrue(metaFile.getFile().exists());
    }

    /**
     * Creates a new {@link MetaFile} for this measurement.
     *
     * @throws IllegalStateException when the {@link Measurement} is in a different state then
     *             {@link Measurement.MeasurementStatus#OPEN}
     * @param vehicle The {@link Vehicle} used in the measurement
     */
    public void createMetaFile(@NonNull final Vehicle vehicle) {
        Validate.isTrue(getStatus() == MeasurementStatus.OPEN, "Unsupported");
        this.metaFile = new MetaFile(this, vehicle);
        Validate.isTrue(metaFile.getFile().exists());
    }

    /**
     * Status which defines weather a measurement is still capturing data, paused, finished, synced or corrupted.
     *
     * @author Armin Schnabel
     * @version 1.0.0
     * @since 3.0.0
     */
    public enum MeasurementStatus {
        OPEN, PAUSED, FINISHED, SYNCED, CORRUPTED
    }
}

File: persistence/src/main/java/de/cyface/persistence/model/Measurement.kt
/*
 * Copyright 2023 Cyface GmbH
 *
 * This file is part of the Cyface SDK for Android.
 *
 * The Cyface SDK for Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The Cyface SDK for Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the Cyface SDK for Android. If not, see <http://www.gnu.org/licenses/>.
 */
package de.cyface.persistence.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import java.util.Objects

/**
 * An `@Entity` which represents a persisted [Measurement] captured by the `DataCapturingService`.
 *
 * An instance of this class represents one row in a database table containing the measurement data.
 *
 * @author Armin Schnabel
 * @version 1.0.0
 * @since 7.5.0
 * @property uid The system wide unique identifier of this measurement. If the entry was not saved
 * yet it may be `null`, since the persistence layers assigns a unique identifier on saving an entry.
 * @property status The capturing status of the measurement, i.e. whether the capturing is still ongoing.
 * @property modality The modality selected when starting this measurement.
 * @property fileFormatVersion The [de.cyface.persistence.PersistenceLayer.PERSISTENCE_FILE_FORMAT_VERSION]
 *           used to serialize the data in the file-based persistence layer of this measurement.
 * @property distance The distance this measurement has captured in meters.
 * @property timestamp The Unix timestamp in milliseconds indicating the start time of the measurement.
 */
@Entity//(tableName = "Measurement")
data class Measurement( // FIXME: make all models `data class`
    @field:PrimaryKey(autoGenerate = true) var uid : Long = 0, // FIXME: Move this to an abstract class "DatabaseEntry" like "DataPoint"
    val status: MeasurementStatus,
    val modality: Modality,
    val fileFormatVersion: Short,
    val distance: Double,
    val timestamp: Long
) /*: Measurement*/ {

    /**
     * Creates a new instance of this class which was not yet persisted and has [uid] set to null. // FIXME: can this also be null?
     *
     * @param status The capturing status of the measurement, i.e. whether the capturing is still ongoing.
     * @param modality The modality selected when starting this measurement.
     * @param fileFormatVersion The [de.cyface.persistence.PersistenceLayer.PERSISTENCE_FILE_FORMAT_VERSION]
     *           used to serialize the data in the file-based persistence layer of this measurement.
     * @param distance The distance this measurement has captured in meters.
     * @param timestamp The Unix timestamp in milliseconds indicating the start time of the measurement.
     */
    constructor(
        status: MeasurementStatus, modality: Modality, fileFormatVersion: Short, distance: Double, timestamp: Long
    ) : this(0, status, modality, fileFormatVersion, distance, timestamp)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Measurement

        if (uid != other.uid) return false
        if (status != other.status) return false
        if (modality != other.modality) return false
        if (fileFormatVersion != other.fileFormatVersion) return false
        if (distance != other.distance) return false
        if (timestamp != other.timestamp) return false

        return true
    }

    override fun hashCode(): Int {
        return Objects.hash(uid)
    }
}